lib/src/working_copy.rs

```diff
@@ -552,9 +572,13 @@ impl TreeState {
             self.file_states.remove(file);
             tree_builder.remove(file.clone());
         }
-        let changed = tree_builder.has_overrides();
-        self.tree_id = tree_builder.write_tree();
-        Ok(changed)
+        let has_changes = tree_builder.has_overrides();
+        let tree_id = tree_builder.write_tree();
+        self.tree_id = tree_id.clone();
+        Ok(Snapshot {
+            has_changes,
+            tree_id,
+        })
```

Comment:
Why is it better to return the struct than to let the caller call `current_tree_id()` if they want the tree id?

Reply:
@martinvonz I think it's a broader philosophical concern. This type has lots of side effects in functions and it's not clear which functions update what state or which functions have to be called in a certain order. There's also the problem of transactional safety.

- For example, one of the previous lines updates `self.watchman_clock`, and I can't easily tell from looking at the code if it's safe to actually do that.
- I also don't know whether updating `file_states` in-place (as done currently) is safe, either: it's possible that `update_file_state` for some file will raise an error, but the already-updated `file_states` will be preserved, so the `TreeState` is invalid and must be discarded (IIUC).

Ideally, `snapshot` wouldn't take `&mut self` at all, and simply return the `Snapshot`, and a separate method would actually commit it. That way, the transactional guarantees are much more obvious and it's easier to commit/discard transactions.

Reply:
That makes sense. We' have to be a bit careful not to make it slower by copying the `TreeState` unnecessarily. Maybe we can have `Snapshot` contain just the changes compared to the base `TreeState` or something.

I'd prefer to leave the introduction of the `Snapshot` for another PR focused on making the interface safer to use. It doesn't seem very relevant to this PR.